pcl_ros/include/pcl_ros/filters/filter.hpp

```diff
@@ -52,19 +52,24 @@ namespace sync_policies = message_filters::sync_policies;
   * applicable to all filters are defined here as static methods.
   * \author Radu Bogdan Rusu
   */
-class Filter : public PCLNodelet
+class Filter : public PCLNode<sensor_msgs::msg::PointCloud2>
 {
 public:
-  typedef sensor_msgs::PointCloud2 PointCloud2;
+  typedef sensor_msgs::msg::PointCloud2 PointCloud2;
 
   typedef pcl::IndicesPtr IndicesPtr;
   typedef pcl::IndicesConstPtr IndicesConstPtr;
 
-  Filter() {}
+  /** \brief Filter constructor
+    * \param node_name node name
+    * \param options node options
+    * \param use_frame_params declare and handle parameter callback for input_frame and output_frame if true
+    */
+  Filter(std::string node_name, const rclcpp::NodeOptions & options, bool use_frame_params=false);
```

Comment:
To be a composable node, the only constructor fields can be the NodeOptions -- is this a node to be deployed itself? If so, its probably wise to use that so that all of these can be composed into a single process to be able to leverage IPC optimizations

Reply:
If not, this is only a base node, that's fine, but just an FYI that this could make some odd looking derived nodes if they have to call the constructor `Filter` in their constructors with hardcoded values of `use_frame_params` (since the others are passed `NodeOptions` or can be hardcoded `node_name`)

Reply:
> is this a node to be deployed itself?

No, this is an abstract class and not deployed.

> If not, this is only a base node, that's fine, but just an FYI that this could make some odd looking derived nodes if they have to call the constructor `Filter` in their constructors with hardcoded values of `use_frame_params` (since the others are passed `NodeOptions` or can be hardcoded `node_name`)

The class has a callback function that handles `dynamic_reconfigure` parameter change for `input_frame` and `output_frame`, however all sub classes do not use this callback and just copy the logic ([comment](https://github.com/ros-perception/perception_pcl/blob/melodic-devel/pcl_ros/src/pcl_ros/filters/voxel_grid.cpp#L119) in the original code).

`use_frame_params` is a replacement of `has_service` flag in `child_init` function, which is not actually worked well in the original code, because all sub class set it to true or override `child_init` itself.

Reply:
But should `use_frame_params` be a constructor argument?